### In this first example, we will explain the functionality of the MOMRollRateTable class

### MOMRollRateTable

In Application and Behavioural Scorecards, most of the time we don't have an exact definition on the bad customers.
So what we need to do is to decide who they are. Most of the time we decide that based on their delinquency status (e.g. 1-month delinquent, 2-month deliqnuent, etc) from month i to month i+1 (suppose September to October).

Firstly, let's see our data. In the tests folder there is a directory simulation_data. We pick 2 files representing month i and month i+1

In [1]:
import polars as pl

data_i = pl.scan_csv("../tests/simulation_data/test_sample_0.csv").collect()
data_i_1 = pl.scan_csv("../tests/simulation_data/test_sample_1.csv").collect()

In [2]:
data_i.head()

shape: (5, 8)
┌──────────────────────────────────┬──────┬──────┬────────┬──────────┬────────┬───────────┬───────────┐
│ id                               ┆ delq ┆ Open ┆ Active ┆ Deactive ┆ Closed ┆ bin_ind_1 ┆ bin_ind_2 │
│ ---                              ┆ ---  ┆ ---  ┆ ---    ┆ ---      ┆ ---    ┆ ---       ┆ ---       │
│ str                              ┆ i64  ┆ i64  ┆ i64    ┆ i64      ┆ i64    ┆ i64       ┆ i64       │
╞══════════════════════════════════╪══════╪══════╪════════╪══════════╪════════╪═══════════╪═══════════╡
│ 950165FF0FFE8826C147C35805AEEAD9 ┆ 0    ┆ 1    ┆ 1      ┆ 0        ┆ 0      ┆ 0         ┆ 1         │
│ 885857A5D4552651CEC59F6897EF2F0B ┆ 0    ┆ 1    ┆ 1      ┆ 0        ┆ 0      ┆ 0         ┆ 1         │
│ CC2672DFA2988C41F3D13A8B205A127E ┆ 0    ┆ 1    ┆ 1      ┆ 0        ┆ 0      ┆ 1         ┆ 0         │
│ D73B2D86E08D8E48238B59A55EAD0CA4 ┆ 0    ┆ 1    ┆ 1      ┆ 0        ┆ 0      ┆ 1         ┆ 0         │
│ 23D46CDE09B973C462C8B7F287F87495 ┆ 1    ┆ 1    ┆ 1      ┆ 0        ┆ 0      ┆ 1         ┆ 0         │
└──────────────────────────────────┴──────┴──────┴────────┴──────────┴────────┴───────────┴───────────┘

In [3]:
data_i_1.head()

shape: (5, 8)
┌──────────────────────────────────┬──────┬──────┬────────┬──────────┬────────┬───────────┬───────────┐
│ id                               ┆ delq ┆ Open ┆ Active ┆ Deactive ┆ Closed ┆ bin_ind_1 ┆ bin_ind_2 │
│ ---                              ┆ ---  ┆ ---  ┆ ---    ┆ ---      ┆ ---    ┆ ---       ┆ ---       │
│ str                              ┆ i64  ┆ i64  ┆ i64    ┆ i64      ┆ i64    ┆ i64       ┆ i64       │
╞══════════════════════════════════╪══════╪══════╪════════╪══════════╪════════╪═══════════╪═══════════╡
│ B1BF9E2890A8D8727E7A3AE81032A434 ┆ 1    ┆ 1    ┆ 1      ┆ 0        ┆ 0      ┆ 1         ┆ 0         │
│ 6EDAD278689D5EC709C1FD3CAEB05D13 ┆ 1    ┆ 1    ┆ 1      ┆ 0        ┆ 0      ┆ 1         ┆ 1         │
│ 8779CF51DFA1152461850FF6D5AF3456 ┆ 1    ┆ 1    ┆ 1      ┆ 0        ┆ 0      ┆ 1         ┆ 0         │
│ 81EA835BB1B98EBCACCAF1E0687A91D0 ┆ 0    ┆ 1    ┆ 1      ┆ 0        ┆ 0      ┆ 0         ┆ 0         │
│ 691C4F9084FC3A553ABE00076AFC39BA ┆ 0    ┆ 1    ┆ 1      ┆ 0        ┆ 0      ┆ 1         ┆ 1         │
└──────────────────────────────────┴──────┴──────┴────────┴──────────┴────────┴───────────┴───────────┘

We can see that there are 8 columns:

**id**: Account id, the unique key of the dataset. \
**delq**: The delinquency of each account. \
**bin_ind_1**: A binary indicator. \
**bin_ind_2**: A binary indicator. \
**Open, Active, Deactive, Closed**: Indicates if the account is open, active, deactive or closed in that month. 

The last 6 columns are not of any use to us for this example.

So in order to get the roll rates from test_sample_0 to test_sample_1 we instanciate a MOMRllRateTable object and passing in the arguments needed, the unique key column of the 2 files **(should have the same name)**, the column which indicates the delinquency **(should have the same name)**, the paths to the 2 files (path_i for month i and path_i_1 for month i+1), the max deliqnuency we want to track and finally if we want to inlcude any binary indicators.

Then we call mehtod build() on the object to calculates the roll rate table.

In [4]:
from roll_rate_analysis import MOMRollRateTable

table = MOMRollRateTable(
    unique_key_col="id",
    delinquency_col="delq",
    path_i="../tests/simulation_data/test_sample_0.csv",
    path_i_1="../tests/simulation_data/test_sample_1.csv",
    max_delq=6,
)

In [5]:
table.build()

In [6]:
table.get_roll_rates()

0_cycle_deliqnuent  1_cycle_deliqnuent  \
0_cycle_deliqnuent                29173                6245   
1_cycle_deliqnuent                 7664                9523   
2_cycle_deliqnuent                 1015                2345   
3_cycle_deliqnuent                  259                 361   
4_cycle_deliqnuent                   61                  35   
5_cycle_deliqnuent                   31                  14   
6+_cycle_deliqnuent                  21                   2   

                     2_cycle_deliqnuent  3_cycle_deliqnuent  \
0_cycle_deliqnuent                    3                   0   
1_cycle_deliqnuent                 2304                   0   
2_cycle_deliqnuent                  352                 719   
3_cycle_deliqnuent                   64                  23   
4_cycle_deliqnuent                   10                   2   
5_cycle_deliqnuent                    0                   2   
6+_cycle_deliqnuent                   1                   1   

                     4_cycle_deliqnuent  5_cycle_deliqnuent  \
0_cycle_deliqnuent                    0                   0   
1_cycle_deliqnuent                    0                   0   
2_cycle_deliqnuent                    0                   0   
3_cycle_deliqnuent                  476                   0   
4_cycle_deliqnuent                   23                 240   
5_cycle_deliqnuent                    5                  13   
6+_cycle_deliqnuent                   3                   3   

                     6+_cycle_deliqnuent  
0_cycle_deliqnuent                     2  
1_cycle_deliqnuent                     1  
2_cycle_deliqnuent                     0  
3_cycle_deliqnuent                     0  
4_cycle_deliqnuent                     0  
5_cycle_deliqnuent                   170  
6+_cycle_deliqnuent                  380

So that does that table tell us? The sum of accounts in row **n** indicates the number of accounts that were **n** cycle delinquent in month i. The sum of accounts in column **k** indicates the number of accounts that were **k** cycle delinquent in month i+1. So the value at **(2,3)** shows that **2304** accounts that were 1 cycle delinquent on month i, are 2 cycle delinquent on month i+1, which means they didn't pay the installment in time.

Then we can call reduce to get a more high level view of that table like below.

In [7]:
table.reduce()

roll_down  stable  roll_up
0_cycle_deliqnuent         0.0    82.4     17.6
1_cycle_deliqnuent        39.3    48.9     11.8
2_cycle_deliqnuent        75.8     7.9     16.2
3_cycle_deliqnuent        57.8     1.9     40.2
4_cycle_deliqnuent        29.1     6.2     64.7
5_cycle_deliqnuent        22.1     5.5     72.3
6+_cycle_deliqnuent        7.5    92.5      0.0

The table above gives us higher level info about the perventages of accounts in each bucket. \
**roll_down**: they paid their installment and at least one previous installment they owed \
**stable**: they just paid current installment (or remained in the 6+ deliqnuency bucket) \
**roll_up**: they didn't pay their current installment and their status chenged to worse

In most cases, the bucket where we see that the roll_up percentage is higher than 50% is the point were we decide **accounts from that bucket and below classify as bad accounts** (i.e. 4, 5, 6+ in our case).

In general, we compute the roll rates for a larger period of time (e.g. a year) and sum them up and go on from that. But for the purpose of this example its not necessary.

# Miscellaneous

If we wanted to track delinquencies bigger than 6 then we could change the max_delq argument at initialization:

In [8]:
table = MOMRollRateTable(
    unique_key_col="id",
    delinquency_col="delq",
    path_i="../tests/simulation_data/test_sample_0.csv",
    path_i_1="../tests/simulation_data/test_sample_1.csv",
    max_delq=7,
)
table.build()
table.get_roll_rates()

0_cycle_deliqnuent  1_cycle_deliqnuent  \
0_cycle_deliqnuent                29173                6245   
1_cycle_deliqnuent                 7664                9523   
2_cycle_deliqnuent                 1015                2345   
3_cycle_deliqnuent                  259                 361   
4_cycle_deliqnuent                   61                  35   
5_cycle_deliqnuent                   31                  14   
6_cycle_deliqnuent                   10                   2   
7+_cycle_deliqnuent                  11                   0   

                     2_cycle_deliqnuent  3_cycle_deliqnuent  \
0_cycle_deliqnuent                    3                   0   
1_cycle_deliqnuent                 2304                   0   
2_cycle_deliqnuent                  352                 719   
3_cycle_deliqnuent                   64                  23   
4_cycle_deliqnuent                   10                   2   
5_cycle_deliqnuent                    0                   2   
6_cycle_deliqnuent                    1                   1   
7+_cycle_deliqnuent                   0                   0   

                     4_cycle_deliqnuent  5_cycle_deliqnuent  \
0_cycle_deliqnuent                    0                   0   
1_cycle_deliqnuent                    0                   0   
2_cycle_deliqnuent                    0                   0   
3_cycle_deliqnuent                  476                   0   
4_cycle_deliqnuent                   23                 240   
5_cycle_deliqnuent                    5                  13   
6_cycle_deliqnuent                    2                   3   
7+_cycle_deliqnuent                   1                   0   

                     6_cycle_deliqnuent  7+_cycle_deliqnuent  
0_cycle_deliqnuent                    0                    2  
1_cycle_deliqnuent                    0                    1  
2_cycle_deliqnuent                    0                    0  
3_cycle_deliqnuent                    0                    0  
4_cycle_deliqnuent                    0                    0  
5_cycle_deliqnuent                  170                    0  
6_cycle_deliqnuent                    3                   65  
7+_cycle_deliqnuent                   1                  311

In [9]:
table.reduce()

roll_down  stable  roll_up
0_cycle_deliqnuent         0.0    82.4     17.6
1_cycle_deliqnuent        39.3    48.9     11.8
2_cycle_deliqnuent        75.8     7.9     16.2
3_cycle_deliqnuent        57.8     1.9     40.2
4_cycle_deliqnuent        29.1     6.2     64.7
5_cycle_deliqnuent        22.1     5.5     72.3
6_cycle_deliqnuent        21.8     3.4     74.7
7+_cycle_deliqnuent        4.0    96.0      0.0